util/intern.cc

```diff
@@ -0,0 +1,71 @@
+#include "intern.h"
+#include "xxhash.h"
+#include <cassert>
+#include <cstring>
+#include <mutex>
+#include <unordered_map>
+
+namespace atlas {
+namespace util {
+
+struct Hasher {
+  size_t operator()(const char* str) const {
+    return XXH64(str, strlen(str), 42);
+  }
+};
+
+struct Comparer {
+  bool operator()(const char* s1, const char* s2) const {
+    return std::strcmp(s1, s2) == 0;
+  }
+};
+
+class StringPool {
+ public:
+  StringPool() noexcept {}
+  ~StringPool() noexcept = default;
+  StringPool(const StringPool& pool) = delete;
+  StringPool& operator=(const StringPool& pool) = delete;
+
+  const StrRef& intern(const char* string) {
+    std::lock_guard<std::mutex> guard(mutex);
+    auto it = table.find(string);
+    if (it != table.end()) {
+      return it->second;
+    }
+    const auto* copy = strdup(string);
+    StrRef ref;
+    ref.data = copy;
+    table.insert(std::make_pair(copy, ref));
+    return table.at(copy);
+  }
+
+  const char* to_string(const StrRef& ref) { return ref.data; }
+
+ private:
+  std::unordered_map<const char*, StrRef, Hasher, Comparer> table;
+
+  std::mutex mutex;
+};
+
+static StringPool the_pool;
```

Comment:
Consider adding a metric to track the pool size.

Reply:
Yes. I definitely want that, plus keep track of them sum of the size of the strings added to the pool. I was going to do that in a separate PR.